recOrder/calib/Calibration.py

```diff
@@ -907,39 +1020,193 @@ def capture_bg(self, n_avg, directory):
+            Retardance in nanometers
+        a, b1, b2, c, d, e : float
+            Fit parameters
+        wavelength : float
+            Wavelength in nanometers
+
+        Returns
+        -------
+        voltage : float
+            Voltage in volts
+
+        """
+
+        voltage = c * (((b1 + b2 / wavelength ** 2) / (retardance - a)) ** (1 / e) - 1) ** (1 / d)
+
+        return voltage
+
+    @staticmethod
+    def _fun(x, wavelengths, xdata, ydata):
+        fval = CalibrationData.schnoor_fit(xdata, *x, wavelengths)
+        res = ydata - fval
+        return res.flatten()
 
     def set_wavelength(self, wavelength):
-        self.wavelength = wavelength
+        if len(self.calib_wavelengths) == 1 and wavelength != self.calib_wavelengths:
+            raise ValueError("Calibration is not provided at this wavelength. "
+                             "Wavelength dependence of LC retardance vs voltage cannot be extrapolated.")
 
-        # Interpolation of curves beyond this range produce strange results.
-        if self.wavelength < 450:
-            self.wavelength = 450
-        if self.wavelength > 720:
-            self.wavelength = 720
+        if wavelength < self.calib_wavelengths.min() or \
+                wavelength > self.calib_wavelengths.max():
+            warnings.warn("Specified wavelength is outside of the calibration range. "
+                          "LC retardance vs voltage data will be extrapolated at this wavelength.")
 
-        self.create_wavelength_curve()
+        self.wavelength = wavelength
+        if self.interp_method == 'linear':
+            # Interpolation of calib beyond this range produce strange results.
+            if self.wavelength < 450:
+                self.wavelength = 450
+                warnings.warn("Wavelength is limited to 450-720 nm for this interpolation method.")
+            if self.wavelength > 720:
+                self.wavelength = 720
+                warnings.warn("Wavelength is limited to 450-720 nm for this interpolation method.")
+
+    def fit_data(self, raw_data, calib_wavelengths):
+        xdata = raw_data[:, 0::3] / 1000    # convert to volts
+        ydata = raw_data[:, 1::3]           # in nanometers
+
+        x0 = [10, 1000, 1e7, 1, 10, 0.1]
+        p = least_squares(self._fun, x0, method='trf', args=(calib_wavelengths, xdata, ydata),
+                          bounds=((-np.inf, 0, 0, 0, 0, 0), (np.inf,)*6),
+                          x_scale=[10, 1000, 1e7, 1, 10, 0.1])
+
+        if not p.success:
+            raise RuntimeError("Schnoor fit to calibration data did not work.")
+
+        y = ydata.flatten()
+        y_hat = y - p.fun
+        slope, intercept, r_value, *_ = linregress(y, y_hat)
+        r_squared = r_value**2
+        if r_squared < 0.999:
+            warnings.warn(f'Schnoor fit has R2 value of {r_squared:.5f}, fit may not have worked well.')
+
+        return p.x
+
+    def interpolate_data(self, raw_data, calib_wavelengths):
```

Comment:
`calib_wavelengths` is unused. I'm guessing that you you were thinking of rewriting in a more generalizable way (instead of a hardcoded small set of wavelengths)? I'm fine with skipping work and just deleting this option since `schnoor_fit` is performed well.

Reply:
Yeah, I was hoping to make the linear fit a bit more generic, but I figures it's not worth the time at this point. I think I kept it in just to mimic the signature of the `fit_data` method. Maybe we can keep it in and just leave a comment in the code for now?

Reply:
Works for me!